examples/vnet/doca_gw_dpdk.c

```diff
@@ -421,6 +423,239 @@ static int doca_gw_dpdk_build_item(struct doca_gw_match *match,
+}
+
+static int doca_gw_dpdk_modify_ipv4_addr_action(struct doca_dpdk_action_entry *entry,
+									struct doca_gw_actions *pkt_action)
+{
+	struct doca_ip_addr *ip_addr;
+	struct rte_flow_action *action = entry->action;
+	struct rte_flow_action_set_ipv4 *ipv4 = &entry->action_data.ipv4.ipv4;
+
+	ip_addr = action->type == RTE_FLOW_ACTION_TYPE_SET_MAC_SRC?
+		&pkt_action->mod_src_ip : &pkt_action->mod_dst_ip;
+	if (!doca_is_ip_zero(ip_addr))
+		ipv4->ipv4_addr = ip_addr->a.ipv4_addr;
+	return 0;
+}
+
+static void doca_gw_dpdk_build_ipv4_addr_action(struct doca_dpdk_action_entry *entry,
+									struct doca_gw_pipeline_cfg *cfg, uint8_t type)
+{
+	struct rte_flow_action *action = entry->action;
+	struct doca_ip_addr *ip_addr;
+	struct rte_flow_action_set_ipv4 *ipv4 = &entry->action_data.ipv4.ipv4;
+
+	ip_addr = (type == DOCA_SRC? &cfg->actions->mod_src_ip : &cfg->actions->mod_dst_ip);
+	ipv4->ipv4_addr = ip_addr->a.ipv4_addr;
+	action->type = (type == DOCA_SRC? RTE_FLOW_ACTION_TYPE_SET_IPV4_SRC : RTE_FLOW_ACTION_TYPE_SET_IPV4_DST);
+	action->conf = &ipv4;
+	if (doca_is_ip_max(ip_addr))
+		entry->modify_action = doca_gw_dpdk_modify_ipv4_addr_action;
+}
+
+static int doca_gw_dpdk_modify_l4_port_action(struct doca_dpdk_action_entry *entry,
+									struct doca_gw_actions *pkt_action)
+{
+	uint16_t l4port;
+	struct rte_flow_action *action = entry->action;
+	struct rte_flow_action_set_tp *set_tp = &entry->action_data.l4port.l4port;
+
+	l4port = action->type == RTE_FLOW_ACTION_TYPE_SET_TP_SRC?
+		pkt_action->mod_src_port : pkt_action->mod_dst_port;
+	if (l4port)
+		set_tp->port = l4port;
+	return 0;
+}
+
+static void doca_gw_dpdk_build_l4_port_action(struct doca_dpdk_action_entry *entry,
+									struct doca_gw_pipeline_cfg *cfg, uint8_t type)
+{
+	uint16_t l4port;
+	struct rte_flow_action *action = entry->action;
+	struct rte_flow_action_set_tp *set_tp = &entry->action_data.l4port.l4port;
+
+	l4port = (type == DOCA_SRC? cfg->actions->mod_src_port : cfg->actions->mod_dst_port);
+	set_tp->port = l4port;
+	action->type = (type == DOCA_SRC? RTE_FLOW_ACTION_TYPE_SET_TP_SRC : RTE_FLOW_ACTION_TYPE_SET_TP_DST);
+	action->conf = &set_tp;
+	if (l4port == UINT16_MAX)
+		entry->modify_action = doca_gw_dpdk_modify_l4_port_action;
+}
+
+static void doca_gw_dpdk_build_dec_ttl_action(struct doca_dpdk_action_entry *entry)
+{
+	struct rte_flow_action *action = entry->action;
+
+	action->type = RTE_FLOW_ACTION_TYPE_DEC_TTL;
+	action->conf = NULL;
+}
+
+static int doca_gw_dpdk_build_action(struct doca_gw_pipeline_cfg *cfg,
+					struct doca_gw_pipe_dpdk_flow *pipe_flow)
+{
+#define NEXT_ACTION (&pipe_flow->action_entry[idx++])
+	int ret = 0;
+	uint8_t idx = 0;
+	struct doca_gw_match *match = cfg->match;
+	struct doca_gw_actions *actions = cfg->actions;
+
+	if (actions->decap && !match->tun.type != DOCA_TUN_NONE) {
+		ret = doca_gw_dpdk_build_tunnel_action(NEXT_ACTION, cfg);
+	}
+	if (!doca_is_mac_zero(actions->mod_src_mac))
```

Comment:
I think you can take same approach as in match. flag what should be modified and what should be const and what you don't care at all.